### PR TITLE
[TT-5805] Update Circuit Breaker triggers

### PR DESCRIPTION
**Pull request by marksou**
Hope I've understood this correctly :-)

### CODE REVIEW

**Comment by andyo-tyk** [comment]
Almost.... :)

1) You need to add the `BreakerTriggered` event to the Event Type list at /docs/basic-config-and-security/report-monitor-trigger-events/event-types/
2) The system will actually create both the BreakerTriggered event and one of the new events (BreakerTripped or BreakerReset); these new events are the ones to which you can attach a webhook event in the dashboard. I think this needs some explanation...

- The `BreakerTriggered` event will come with a status code that tells you which event it was (tripped or reset)
- If the user wants to use the dashboard to connect their webhooks, then they must use the `BreakerTripped` and `BreakerReset` events - the dashboard does not use `BreakerTriggered`
- I'm wondering whether the new screenshot should be at the bottom of the page, as you do that bit last (after configuring the webhook, you have to attach it to the correct event)?

**Comment by marksou** [comment]
Hi @andyo-tyk I've pushed some further changes after your comments.

File: tyk-docs/content/basic-config-and-security/report-monitor-trigger-events/event-types.md [hunk 1]
```diff
@@ -20,8 +20,9 @@ The events currently raised by Tyk are:
 *   `AuthFailure`: A key has failed authentication or has attempted access and was denied
 *   `KeyExpired`: A key has attempted access but is expired
 *   `VersionFailure`: A key has attempted access to a version it does not have permission to access
-*   `BreakerTripped`: When a circuit breaker on a path trips and a service is taken offline
-*   `BreakerReset`: When the circuit breaker comes back on-stream
+*   `BreakerTriggered`: When a circuit breaker trips, this event will be created along with either a `BreakerTripped`, or a `BreakerReset` event. A status code will indicate which of these events was triggered. This event is not supported in the Tyk Dashboard.
```
Comment (andyo-tyk) [inline tyk-docs/content/basic-config-and-security/report-monitor-trigger-events/event-types.md]: *   `BreakerTriggered`: This event will be created when either a `BreakerTripped`, or a `BreakerReset` event occurs; a status code in the metadata passed to the webhook will indicate which of these events was triggered. This event is not supported in the Tyk Dashboard, but is provided as an alternative to allow configuration of a single webhook for the two circuit breaker events.

File: tyk-docs/content/basic-config-and-security/report-monitor-trigger-events/event-types.md [hunk 2]
```diff
@@ -20,8 +20,9 @@ The events currently raised by Tyk are:
...
+*   `BreakerTripped`: When a circuit breaker on a path trips and a service is taken offline. This can be set as a webhook event in the Dashboard
```
Comment (andyo-tyk) [inline tyk-docs/content/basic-config-and-security/report-monitor-trigger-events/event-types.md]: Probably don't need to add the second sentence, with my suggested rewording of BreakerTriggered - none of the other events state that they can be set in the Dashboard, so it looks a bit inconsistent.

File: tyk-docs/content/basic-config-and-security/report-monitor-trigger-events/event-types.md [hunk 3]
```diff
@@ -20,8 +20,9 @@ The events currently raised by Tyk are:
...
+*   `BreakerReset`: When the circuit breaker comes back on-stream. This can be set as a webhook event in the Dashboard
```
Comment (andyo-tyk) [inline tyk-docs/content/basic-config-and-security/report-monitor-trigger-events/event-types.md]: Probably don't need to add the second sentence, with my suggested rewording of BreakerTriggered - none of the other events state that they can be set in the Dashboard, so it looks a bit inconsistent.

File: tyk-docs/content/planning-for-production/ensure-high-availability/circuit-breakers.md [hunk 4]
```diff
@@ -98,3 +105,7 @@ Once the plugin is active, you can set up the various configurations options for
 *   **Trigger threshold percentage**: The percentage of requests that can error before the breaker is tripped, this must be a value between 0.0 and 1.0.
 *   **Sample size (requests)**: The number of samples to take for a circuit breaker window.
 *   **Return to service in (s)**: The cool-down period of the breaker to return to service (seconds).
+
+In the Dashboard, the `BreakerTrigger` [event type]({{< ref "/content/basic-config-and-security/report-monitor-trigger-events/event-types.md" >}}) is not supported. Instead, you need to add a Webhook plugin from the **API Designer > Advanced Options** to your endpoint and select either the `BreakerTripped` or `BreakerReset` events.
```
Comment (andyo-tyk) [inline tyk-docs/content/planning-for-production/ensure-high-availability/circuit-breakers.md]:
```suggestion
The Dashboard supports the separate `BreakerTripped` and `BreakerReset` events, but not the combined `BreakerTrigger` [event type]({{< ref "/content/basic-config-and-security/report-monitor-trigger-events/event-types.md" >}}). You should use **API Designer > Advanced Options** to add a Webhook plugin to your endpoint for each event.
```

File: tyk-docs/content/planning-for-production/ensure-high-availability/circuit-breakers.md [hunk 5]
```diff
@@ -13,15 +13,21 @@ weight: 3
...
 
 #### Events
 
-When a circuit breaker trips, it will fire a `BreakerTriggered` event which you can define actions for in the `event_handlers` section (see [Event Data](/docs/basic-config-and-security/report-monitor-trigger-events/event-data/) and [Event Types](/docs/basic-config-and-security/report-monitor-trigger-events/event-types/) for more information):
+When a circuit breaker trips, it can fire a `BreakerTriggered` [event type]({{< ref "/content/basic-config-and-security/report-monitor-trigger-events/event-types.md" >}}) which you can define actions for in the `event_handlers` section (see [Event Data](/docs/basic-config-and-security/report-monitor-trigger-events/event-data/) and [Event Types](/docs/basic-config-and-security/report-monitor-trigger-events/event-types/) for more information).
+
+{{< note success >}}
+**Note**  
+
+The `BreakerTriggered` event cannot be configured via the Tyk Dashboard and you need to use Webhooks to specify the `BreakerTripped` and `BreakerReset` event. See [Configure with the Dashboard](#configure-with-the-dashboard) for more details.
```
Comment (andyo-tyk) [inline tyk-docs/content/planning-for-production/ensure-high-availability/circuit-breakers.md]:
```suggestion
The Dashboard supports the separate `BreakerTripped` and `BreakerReset` events, but not the combined `BreakerTrigger` event. See [Configure with the Dashboard](#configure-with-the-dashboard) for more details.
```

File: tyk-docs/content/planning-for-production/ensure-high-availability/circuit-breakers.md [hunk 6]
```diff
@@ -13,15 +13,21 @@ weight: 3
...
... built-in circuit breaker pattern as a path-based option. Our circuit breaker is rate-based, so if a sample size `x` of `y%` requests fail, the breaker will trip. The Gateway will stop **all** inbound requests to that service for a pre-defined period of time (a recovery time-period). You configure this time period using the `return_to_service_after` option in your API definition, or setup via the Dashboard. The circuit breaker also has an half-open state that can check if the problem is fixed, this is done by making real requests to the upstream before the time configured in  `return_to_service_after` happens. By default the Tyk circuit breaker has enabled the half-open state, if the desired behavior is to only check after the time configured in `return_to_service_after` is consumed then you can disable this by setting  `disable_half_open_state` to `true`. See [Configure with the API Definition](#configure-with-the-api-definition) or [Configure with the Dashboard](#configure-with-the-dashboard). This also triggers an event which you can hook into to perform corrective or logging action. When a circuit breaker is tripped, it will return a 503 "Service temporarily unavailable" error.
```
Comment (andyo-tyk) [inline tyk-docs/content/planning-for-production/ensure-high-availability/circuit-breakers.md]:
```suggestion
Tyk has a built-in circuit breaker pattern as a path-based option. Our circuit breaker is rate-based, so if a sample size `x` of `y%` requests fail, the breaker will trip.  This triggers an event which you can hook into to perform corrective or logging action. 

The Gateway will stop **all** inbound requests to that service for a pre-defined period of time (a recovery time-period). You can configure this recovery time-period using the `return_to_service_after` option in your API definition, or via the Dashboard.

Once a circuit breaker has been tripped, the Tyk Gateway will return a 503 "Service temporarily unavailable" error for any calls to the API until the end of the recovery time-period.

During the recovery time-period, the Tyk Gateway will periodically issue requests to the upstream service to check whether the path has been restored. If the gateway detects that the path has been reconnected, the circuit breaker will be automatically reset and the 'BreakerReset' event will be generated.

This behaviour is described as the circuit breaker being "half-open"; if the desired behaviour is to enforce the full recovery time period (i.e. to unblock the path only after the `return_to_service_after` time period) then you can disable the half-open operation by setting `disable_half_open_state` to `true`.

See [Configure with the API Definition](#configure-with-the-api-definition) or [Configure with the Dashboard](#configure-with-the-dashboard).
```

**Review by andyo-tyk — CHANGES_REQUESTED** [review]
Re-worded the section on half-open state and some suggestions to clarify the different Circuit Breaker triggers.

File: tyk-docs/content/planning-for-production/ensure-high-availability/circuit-breakers.md [hunk 7]
```diff
@@ -13,15 +13,31 @@ weight: 3
...
 
 #### Events
 
-When a circuit breaker trips, it will fire a `BreakerTriggered` event which you can define actions for in the `event_handlers` section (see [Event Data](/docs/basic-config-and-security/report-monitor-trigger-events/event-data/) and [Event Types](/docs/basic-config-and-security/report-monitor-trigger-events/event-types/) for more information):
+When a circuit breaker trips, it can fire a `BreakerTriggered` [event type]({{< ref "/content/basic-config-and-security/report-monitor-trigger-events/event-types.md" >}}) which you can define actions for in the `event_handlers` section (see [Event Data](/docs/basic-config-and-security/report-monitor-trigger-events/event-data/) and [Event Types](/docs/basic-config-and-security/report-monitor-trigger-events/event-types/) for more information).
+
+{{< note success >}}
+**Note**  
+
+The Dashboard supports the separate `BreakerTripped` and `BreakerReset` events, but not the combined `BreakerTrigger` event. See [Configure with the Dashboard](#configure-with-the-dashboard) for more details.
```
Comment (andyo-tyk) [inline tyk-docs/content/planning-for-production/ensure-high-availability/circuit-breakers.md]: Should be `BreakerTriggered` (probably my typo, sorry)

File: tyk-docs/content/planning-for-production/ensure-high-availability/circuit-breakers.md [hunk 8]
```diff
@@ -98,3 +115,7 @@ Once the plugin is active, you can set up the various configurations options for
 *   **Trigger threshold percentage**: The percentage of requests that can error before the breaker is tripped, this must be a value between 0.0 and 1.0.
 *   **Sample size (requests)**: The number of samples to take for a circuit breaker window.
 *   **Return to service in (s)**: The cool-down period of the breaker to return to service (seconds).
+
+The Dashboard supports the separate `BreakerTripped` and `BreakerReset` events, but not the combined `BreakerTrigger` [event type]({{< ref "/content/basic-config-and-security/report-monitor-trigger-events/event-types.md" >}}). You should use **API Designer > Advanced Options** to add a Webhook plugin to your endpoint for each event.
```
Comment (andyo-tyk) [inline tyk-docs/content/planning-for-production/ensure-high-availability/circuit-breakers.md]: Same here - it should be `BreakerTriggered`

File: tyk-docs/content/basic-config-and-security/report-monitor-trigger-events/event-types.md [hunk 9]
```diff
@@ -20,8 +20,9 @@ The events currently raised by Tyk are:
 *   `AuthFailure`: A key has failed authentication or has attempted access and was denied
 *   `KeyExpired`: A key has attempted access but is expired
 *   `VersionFailure`: A key has attempted access to a version it does not have permission to access
-*   `BreakerTripped`: When a circuit breaker on a path trips and a service is taken offline
-*   `BreakerReset`: When the circuit breaker comes back on-stream
+*   `BreakerTriggered`: This event will be created when either a `BreakerTripped`, or a `BreakerReset` event occurs; a status code in the metadata passed to the webhook will indicate which of these events was triggered. This event is not supported in the Tyk Dashboard, but is provided as an alternative to allow configuration of a single webhook for the two circuit breaker events.
```
Comment (andyo-tyk) [inline tyk-docs/content/basic-config-and-security/report-monitor-trigger-events/event-types.md]: I think we can simplify this, as it's now well explained in the Circuit Breaker section, so I now suggest dropping the second sentence (and the full stop).

File: tyk-docs/content/planning-for-production/ensure-high-availability/circuit-breakers.md [hunk 10]
```diff
@@ -13,15 +13,31 @@ weight: 3
...
+Tyk has a built-in circuit breaker pattern as a path-based option. Our circuit breaker is rate-based, so if a sample size `x` of `y%` requests fail, the breaker will trip.  This triggers an event which you can hook into to perform corrective or logging action. 
+
+The Gateway will stop **all** inbound requests to that service for a pre-defined period of time (a recovery time-period). You can configure this recovery time-period using the `return_to_service_after` option in your API definition, or via the Dashboard.
+
+Once a circuit breaker has been tripped, the Tyk Gateway will return a 503 "Service temporarily unavailable" error for any calls to the API until the end of the recovery time-period.
+
+During the recovery time-period, the Tyk Gateway will periodically issue requests to the upstream service to check whether the path has been restored. If the gateway detects that the path has been reconnected, the circuit breaker will be automatically reset and the 'BreakerReset' event will be generated.
```
Comment (andyo-tyk) [inline tyk-docs/content/planning-for-production/ensure-high-availability/circuit-breakers.md]: The formatting hasn't worked on this line for `BreakerReset`

**Review by andyo-tyk — CHANGES_REQUESTED** [review]
Hi Mark,
Sorry, I've spotted a couple of typos (probably in what I proposed) and I think we can now simplify the description in the Event Types list.
Thanks,

**Review by andyo-tyk — APPROVED** [review]
Looks great to me - I'll stop fiddling now. :-)
Thanks Mark.

**Pull request merged**